Hexcurse modernization and maintainance

Bosiux opened the pull request:
# 1. Added terminal label click events
![hexcurse-example](https://github.com/user-attachments/assets/85f05e44-826f-4b88-afd5-726934dc386d)
- [x] Help label
- [x] Save label
- [x] Open label
- [x] Goto label
- [x] Find label
- [x] Hex Addr label
- [x] Hex Edit | ASCII Edit label
- [x] Quit label

# 2. Codebase maintainance for `old-style-definition`
### Before:
```C
int
_getopt_internal (argc, argv, optstring, longopts, longind, long_only)
     int argc;
     char *const *argv;
     const char *optstring;
     const struct option *longopts;
     int *longind;
     int long_only;
{
  int option_index;
  ...
```
### After:
```C
int
_getopt_internal (int argc,char *const *argv,const char *optstring,const struct option *longopts,int *longind,int long_only)
{
  int option_index;
  ...
```

# 3. Data conversion loss prevention
### Before:
```C
void make_delta2(int *delta2, int *pat, size_t patlen) {
    size_t  p;
    size_t  last_prefix_index = patlen-1;

    // first loop
    for (p=patlen-1; ;p--) {
        if (is_prefix(pat, patlen, p+1)) {
            last_prefix_index = p+1;
        }
        delta2[p] = (int) (last_prefix_index + (patlen-1 - p));
        if (p == 0) break;
    }

    // second loop
    for (p=0; p < patlen-1; p++) {
        int slen = suffix_length(pat, patlen, p);
        if (pat[p - slen] != pat[patlen-1 - slen]) {
            delta2[patlen-1 - slen] = (int) (patlen-1 - p + slen);
        }
    }
}
```
### After:
```C
void make_delta2(int *delta2, int *pat, size_t patlen) {
    if (!delta2 || !pat || patlen == 0) {
        return;
    }

    size_t p;
    size_t last_prefix_index = patlen - 1;

    // First loop 
    for (p = patlen - 1; ; p--) {
        if (is_prefix(pat, patlen, p + 1)) {
            last_prefix_index = p + 1;
        }

        if (p < patlen) {
            delta2[p] = (int)(last_prefix_index + (patlen - 1 - p));
        } else {
            continue;
        }

        if (p == 0) break;
    }

    // Second loop 
    for (p = 0; p < patlen - 1; p++) {
        size_t slen = suffix_length(pat, patlen, p);
        if (slen <= p) {
            if (pat[p - slen] != pat[patlen - 1 - slen]) {
                size_t delta_index = patlen - 1 - slen;
                if (delta_index < patlen) {
                    delta2[delta_index] = (int)(patlen - 1 - p + slen);
                } else {
                    continue;
                }
            }
        }
    }
}
```
# 4. Shadow variables avoid
### Before:
```C
off_t gotoLine(FILE *fp, off_t currLoc, off_t gotoLoc, off_t maxlines, WINDOW *win)
```
### After:
```C
off_t gotoLine(FILE *fp, off_t currLoc, off_t gotoLoc, off_t maxDisplaylines, WINDOW *win)
```

Bosiux committed:
Added buttons functionality

Bosiux committed:
Data loss prevention & modernization

Bosiux committed:
Autotools configuration update

prso commented:
@Bosiux thanks for your PR.

I like the mouse functionality but I have modified it using key emulation instead of gotos and labels. This and the other improvements (with duplicated README and configure.ac files removed) are now in the [branch mouse](https://github.com/prso/hexcurse-ng/tree/mouse), if it's ok I will merge into master.

Bosiux commented:
@prso thank you.

I'll work on it in the next days using `uinput` kernel module so it will work for any linux based os.
Some of the modification i made in commits [8d4809e](https://github.com/prso/hexcurse-ng/pull/7/commits/8d4809e74529b121bf83c16a8a0c10027504dff9) and [e5fc0b4](https://github.com/prso/hexcurse-ng/pull/7/commits/e5fc0b420fa11cb2c3077dcd6d5319d6e3572039) are also made to allow me to pack it on RHEL systems, so its easy to install either on debian and fedora.

I've tested a bit with my fork, you can check it on  [copr](https://copr.fedorainfracloud.org/coprs/brosion/hexcurse), once we merge to the upstraem `master` we could pack either as new or same version (1.70).

Bosiux committed:
Goto and Labels switched to key remap

Bosiux commented:
> @prso thank you.
> 
> I'll work on it in the next days using `uinput` kernel module so it will work for any linux based os. Some of the modification i made in commits [8d4809e](https://github.com/prso/hexcurse-ng/pull/7/commits/8d4809e74529b121bf83c16a8a0c10027504dff9) and [e5fc0b4](https://github.com/prso/hexcurse-ng/pull/7/commits/e5fc0b420fa11cb2c3077dcd6d5319d6e3572039) are also made to allow me to pack it on RHEL systems, so its easy to install either on debian and fedora.
> 
> I've tested a bit with my fork, you can check it on [copr](https://copr.fedorainfracloud.org/coprs/brosion/hexcurse), once we merge to the upstraem `master` we could pack either as new or same version (1.70).

## Update [#8f097cd](https://github.com/prso/hexcurse-ng/pull/7/commits/8f097cd0ff2ad94401137ffedfedaddaeea4adb9)

I took literally your 'key emulation' first, but digging even deeper i noticed that many of the possible ways include requiring sudo permissions.
Onestly this does seems a little too much for just a such little feature, maybe I misunderstood your request or something else.
In this commit linked above I did a remap of the keys with the corresponding macro for each case, removing all goto and Labels since its a quite old C feature coming from asm and it might also confuse someone else looking at the codebase in the next years.
Although this isn't an "elegant" way, I tested quite a bit and it works.
Let me know for anything.

Bosiux commented:
> Yes, when I said "key emulation" I really meant "key remapping", sorry for the misunderstanding, so I didn't understand the uinput option. Also remember that hexcurse should be executed in any Unix (I think any uinput solution would be Linux only) provided that ncurses library is available.
> 
> In fact for key remapping I already did something very similar as you but using function-keys codes, see the commit [db45bf6](https://github.com/prso/hexcurse-ng/commit/db45bf684895b756e77ea7ec6ba0266b702599cb) in the [mouse](https://github.com/prso/hexcurse-ng/tree/mouse) branch (I modified your original commit but kept the credits for you).
> 
> ```
>  if (x >= 0 && x <= 7)   /* Help label */
>  {
>      key = KEY_F(1);
>  } else if ( x <= 16 && x >= 9)  /* Save label */
>  {
>      key = KEY_F(2);
>  } else if ( x <= 25 && x >= 18) /* Open label */
>  {
>      key = KEY_F(3);
> ```
> 
> Also I added a correction for labels 4 and 5 when the number of columuns (COLS) is odd: ` int odd_add1 = COLS % 2;`
> 
> ```
>  } else if ( x <= (COLS/2) - 2 + odd_add1 && x >= (COLS/2) - 9 + odd_add1) /* Goto label */
>  {
>      key = KEY_F(4);
>  } else if ( x <= (COLS/2) + 7 + odd_add1 && x >= (COLS/2) + odd_add1) /* Find label */
>  {
>      key = KEY_F(5);
> ```

Oh sorry, i haven't seen your KEY_F() , it's better than my CTRL_AND() so I think we should keep yours.

Before closing this pull if everything is ok i need to ask you something.
I'm not sure about certains file license, could you confirm to me what's written below: 

## These files are licensed under GPL-2.0-or-later:
- include/hex.h
- src/acceptch.c
- src/file.c
- src/hexcurse.c
- src/llist.c
- src/screen.c
- src/stack.c
## These files are licensed under GPL-3.0-or-later:
- config.guess
- config.sub

prso commented:
I think that `config.guess` and `config.sub` were generated by `autotools` (with a version that is GPLv3+) while the rest are the sources for hexcurse itself, are GPLv2+ given by the original authors (I'm not) probably when GPLv3 even didn't exist yet and I didn't modify the license.

But I think it's fine.

Bosiux commented:
> I think that `config.guess` and `config.sub` were generated by `autotools` (with a version that is GPLv3+) while the rest are the sources for hexcurse itself, are GPLv2+ given by the original authors (I'm not) probably when GPLv3 even didn't exist yet and I didn't modify the license.
> 
> But I think it's fine.

I did one last commit because I probably missed out in some others commit a change of permissions, many files has  100755 permission `-rwxrwxrwx`, the only problem is that `x` is not required for anyone, since are not binaryes, and as I got suggested i changed them to  100644 `-rw-rw-rw` and let only directoryes with `x` permission, in order to have access to them.
It's not a crucial commit but its just too keep things in order, probably I messed up first so i fixed it back.

I didt one last build on a couple of systems, you can look up [here](https://copr.fedorainfracloud.org/coprs/brosion/hexcurse/build/8876243/).
It's fine for me, if you would like to apply some changes or so its up to you.

Bosiux commented:
I checked again and I deleted last commit, the permission change was due to too many files with executable permission that weren't executed, if you say its normal we can let it there.

prso commented:
I documented the mouse feature in man and help.

I also discovered a regression when clicking mouse buttons in help and input line, for the first I activated keypad(), for the second I just disabled mouse during input (for save, open, goto and find functions).

If all is fine I will merge these changes in [mouse](https://github.com/prso/hexcurse-ng/tree/mouse) into master.

prso commented:
Merged into master.
